### PR TITLE
[decomp] Support matmul folding for zero-copy non-contiguous views in torch.compile

### LINKED ISSUE #197083: [Inductor] Support matmul folding for zero-copy non-contiguous views in torch.compile
## Summary

In the Inductor Python decomposition, `should_fold` returns `False` for some non-contiguous tensors that can be reshaped without a copy.

## Environment

- PyTorch 2.10.0+cu128
- NVIDIA A100
- Affected code: `torch/_decomp/decompositions.py`, `should_fold`

### Direct predicate

```python
import torch
from torch._decomp.decompositions import should_fold

B, K, N = 4, 8, 5
lhs = torch.empty_strided((B, 1, K), (16, 4, 1)).normal_()
rhs = torch.randn((K, N))
folded = lhs.reshape(B, K)

print(lhs.shape)  # torch.Size([4, 1, 8])
print(lhs.stride())  # (16, 4, 1)
print(lhs.is_contiguous())  # False
print(folded.stride())  # (16, 1)
print(folded.untyped_storage().data_ptr() == lhs.untyped_storage().data_ptr())  # True
print(should_fold(lhs, rhs, False))  # False
```

`lhs.reshape(B, K)` is a view: it has the same storage and produces shape `[4, 8]` with stride `(16, 1)`. The predicate nevertheless returns `False` for this layout.

### End-to-end `torch.compile` example

```python
import torch

B, S, K, N = 384, 8192, 4096, 4096

class MatmulView(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = torch.nn.Parameter(
            torch.randn(2, K, N, device="cuda", dtype=torch.float16)
        )

    def forward(self, x):
        lhs = x[:, :K].unsqueeze(1)  # [B, 1, K], stride [S, K, 1]
        rhs = self.weight[0]  # [K, N], parameter view
        return torch.matmul(lhs, rhs)

x = torch.randn(B, S, device="cuda", dtype=torch.float16)
model = MatmulView().eval().requires_grad_(False)
compiled = torch.compile(model, backend="inductor", fullgraph=True, dynamic=False)

with torch.inference_mode():
    compiled(x)  # warmup/compilation
    torch.cuda.synchronize()
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        record_shapes=True,
        with_stack=True,
    ) as prof:
        compiled(x)
        torch.cuda.synchronize()

print(prof.key_averages().table(sort_by="self_cuda_time_total", row_limit=20))
prof.export_chrome_trace("trace.json")
```

On the tested environment, the profile includes aten::bmm and an additional layout operation such as aten::copy_.

## Observed behavior

For the direct example, `should_fold` returns `False` even though the higher-rank operand can be reshaped to `[B, K]` as a view. In the compiled example, the decomposition follows the batched `bmm` path for the same layout.

## Expected behavior

The Python decomposition could use the same viewability criterion as native `matmul`.If the higher-rank operand can be reshaped to (product(t1.shape[:-1]), t1.shape[-1]) without a copy, it should be eligible for folding to mm or mv. This can remove an unnecessary layout copy and allow large cases to use a single `mm`, with the potential for substantially lower GPU time than a batched `bmm` of many `M=1` matrices. 

## Related links

- Native matmul folding change: [pytorch/pytorch#186178](https://github.com/pytorch/pytorch/pull/186178)
- Original native issue: [pytorch/pytorch#186148](https://github.com/pytorch/pytorch/issues/186148)

### Alternatives

The Python decomposition can use the same viewability criterion as native `matmul`. Allow folding when the tensor can be folded without a copy; lhs does not need to be contiguous.

### Additional context

_No response_

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by ZRICHARD9527**
## Issue

Fixes #197083

## Summary

The Python `matmul` decomposition did not use the same stride check as native `matmul`. It rejected some non-contiguous inputs that can be reshaped without a copy. As a result, `torch.compile` kept the `bmm` path and performed an extra layout copy.

This change uses `_compute_stride` for the view check. It also checks `tensor2.mT` when the right-hand input has more dimensions, matching native `matmul`.

## Test Plan

- Added tests for:
  - zero-copy folding with size-1 dimensions;
  - padded non-contiguous strides;
  - layouts that cannot be folded;
  - the higher-rank right-hand-side path using `mT`.
- `python -m py_compile torch/_decomp/decompositions.py test/test_decomp.py`
- `lintrunner --take FLAKE8,RUFF,PYFMT torch/_decomp/decompositions.py test/test_decomp.py`
- `git diff --check`

The full PyTorch test suite was not run because this source checkout does not have a local PyTorch build.

## Benchmark

Environment:

- PyTorch: `2.10.0+cu128`
- GPU: NVIDIA A100
- dtype: `float16`
- Backend: `torch.compile(inductor)`

The following standalone example reproduces the compiled path. Change `B`, `S`, `K`, and `N` to profile other shapes.  The example below uses `B=384, S=8192, K=N=4096`. `x[:, :K].unsqueeze(1)` creates the non-contiguous input with stride `(S, K, 1)`.

```python
import torch

B, S, K, N = 384, 8192, 4096, 4096

class MatmulView(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = torch.nn.Parameter(
            torch.randn(2, K, N, device="cuda", dtype=torch.float16)
        )

    def forward(self, x):
        lhs = x[:, :K].unsqueeze(1)  # [B, 1, K], stride [S, K, 1]
        rhs = self.weight[0]  # [K, N], parameter view
        return torch.matmul(lhs, rhs)


x = torch.randn(B, S, device="cuda", dtype=torch.float16)
model = MatmulView().eval().requires_grad_(False)
compiled = torch.compile(
    model,
    backend="inductor",
    fullgraph=True,
    dynamic=False,
)

with torch.inference_mode():
    compiled(x)
    torch.cuda.synchronize()
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        record_shapes=True,
        with_stack=True,
    ) as prof:
        compiled(x)
        torch.cuda.synchronize()

print(prof.key_averages().table(sort_by="self_cuda_time_total", row_limit=20))
prof.export_chrome_trace("trace.json")
```

These results come from `torch.profiler`. The CUDA column is the matmul kernel self time. Total CPU time includes compiler and runtime overhead and is shown for reference.

Before the patch:

| Configuration (B/S/K/N) | lhs shape (stride) | bmm CPU time | bmm CUDA kernel | Total CUDA time | Total CPU time |
| --- | --- | ---: | ---: | ---: | ---: |
| 16/128/64/128 | (16, 1, 64) (128, 64, 1) | 404.71 μs | 6.94 μs | 6.94 μs | 3.005 ms |
| 32/1024/512/512 | (32, 1, 512) (1024, 512, 1) | 315.04 μs | 15.71 μs | 15.71 μs | 2.745 ms |
| 64/2048/1024/2048 | (64, 1, 1024) (2048, 1024, 1) | 179.87 μs | 77.98 μs | 77.98 μs | 863.60 μs |
| 128/4096/2048/2048 | (128, 1, 2048) (4096, 2048, 1) | 1.246 ms | 258.37 μs | 258.37 μs | 2.908 ms |
| 384/8192/4096/4096 | (384, 1, 4096) (8192, 4096, 1) | 624.73 μs | 2.925 ms | 2.925 ms | 4.069 ms |
| 512/8192/4096/4096 | (512, 1, 4096) (8192, 4096, 1) | 613.67 μs | 3.897 ms | 3.897 ms | 5.100 ms |

After the patch:

| Configuration (B/S/K/N) | lhs shape (stride) | mm CPU time | mm CUDA kernel | Total CUDA time | Total CPU time |
| --- | --- | ---: | ---: | ---: | ---: |
| 16/128/64/128 | (16, 1, 64) (128, 64, 1) | 223.21 μs | 3.68 μs | 3.68 μs | 2.845 ms |
| 32/1024/512/512 | (32, 1, 512) (1024, 512, 1) | 98.48 μs | 5.41 μs | 5.41 μs | 2.279 ms |
| 64/2048/1024/2048 | (64, 1, 1024) (2048, 1024, 1) | 85.12 μs | 8.29 μs | 8.29 μs | 781.78 μs |
| 128/4096/2048/2048 | (128, 1, 2048) (4096, 2048, 1) | 472.37 μs | 12.54 μs | 12.54 μs | 1.152 ms |
| 384/8192/4096/4096 | (384, 1, 4096) (8192, 4096, 1) | 486.37 μs | 61.28 μs | 61.28 μs | 1.205 ms |
| 512/8192/4096/4096 | (512, 1, 4096) (8192, 4096, 1) | 405.31 μs | 80.26 μs | 80.26 μs | 939.10 μs |

CUDA kernel speedup:

| Configuration (B/S/K/N) | Before: bmm | After: mm | Speedup |
| --- | ---: | ---: | ---: |
| 16/128/64/128 | 6.94 μs | 3.68 μs | 1.9x |
| 32/1024/512/512 | 15.71 μs | 5.41 μs | 2.9x |
| 64/2048/1024/2048 | 77.98 μs | 8.29 μs | 9.4x |
| 128/4096/2048/2048 | 258.37 μs | 12.54 μs | 20.6x |
| 384/8192/4096/4096 | 2.925 ms | 61.28 μs | 47.7x |
| 512/8192/4096/4096 | 3.897 ms | 80.26 μs | 48.6x |

compare

The larger configurations benefit most. The original path runs a batched operation over many `M=1` matrices. The folded path uses one larger `mm` and avoids the input-layout copy.

## Checklist

- [x] Passes lint for modified files
- [x] Added/updated tests
- [ ] Updated documentation (not applicable)
- [x] Included benchmark results

## BC-breaking?

No.

> AI assistance disclosure: AI tools helped draft parts of this PR description and review the implementation. I reviewed the code and test results and take responsibility for the change.

### CODE REVIEW

**Comment by ZRICHARD9527** [comment]
Hi @ezyang @penguinwu @bobrenjc93, this is my first PyTorch contribution. This PR updates the matmul decomposition to support zero-copy folding for non-contiguous views in torch.compile. The CI workflows are currently awaiting maintainer approval. Could you please take a look when you have a chance? Thanks!

**Review by ezyang — APPROVED** [review]

**Comment by ZRICHARD9527** [comment]
@ezyang Hi, thanks for the approval! It seems like this PR needs a release notes label. I'm not sure which one would be appropriate. Could you please advise which release notes label I should use?

**Comment by ezyang** [comment]
test fails look real
